### PR TITLE
Adding a Min Risk trainer 

### PR DESCRIPTION
**Pull request by StoyanVenDimitrov**
Hey, I wanted to add a branch for my work on minimum risk trainer. So far I have the mrt trainer, with some issue on memory consumption. I couldn't train with sufficiently large number of samples because of that. I hope you can help out with that. 
Also I start to work on applying deterministic annealing on that. 

Best, 
Stoyan

### CODE REVIEW

**Comment by jlibovicky** [comment]
Hi, your sampling procedure is not correct. You use the `decoded_logits` for sampling, however the distributions in this list are already conditioned on the word previously selected by the greedy decoding.

To be more specific: let's assume you were lucky and your sampling was identical to greedy decoding for _i_-1 steps. Now, you have sampled word `u` in the _i_-th step, whereas the greedy decoding selected word `v`. The distribution you should sample from in the (_i_+1)-th step should be conditioned on `u`:
```
p(w | u, <decoded until i-1>)
```
If you use the list called `decoded_logits`, you will however sample from distribution
```
p(w | v, <decoded until i-1>)
```
conditioned on `v`, no matter what you have sampled in the _i_-th step.

In this way, you will not end up with a probability distribution over the sentences because the chain rule for conditional probabilities does not apply here.

Also, please note that we have quite strict guidelines for code style and use `pylint`, `pycodestyle` for the style checking. You can find more in documentation.

**Comment by StoyanVenDimitrov** [comment]
Hi, 
using pylint should "clean up" the style, right? But what does Travis CI do, does it run some unit tests? What should I do with my code to past the Travis CI?

**Comment by jindrahelcl** [comment]
Hi, pylint, pycodestyle and mypy just tells you what's wrong with the code. Especially mypy can be quite useful for discovering potential bugs. You can install these packages simply with pip:

```
pip install pylint pycodestyle mypy-lang
```

**Comment by StoyanVenDimitrov** [comment]
Hi,
I need some help with the sampling.
I see, that what you say means, sampling is to be done in the decoder, not in the trainer,right? 
But where is the place where the conditioning on a word already selected by the greedy decoder happens? I have these suggestions:
1. The distribution you are talking about  is obtained in `_decoding_loop` using the `step` function.
2. Sampling the next output means `next_word_id = tf.multinomial(step_logits, 1)`, sampling one sample, instead of `next_word_id = tf.argmax(step_logits, axis=1)`  l.426, decoder.py
But actually don't see `next_word_id` to be reused in the next step.

**Comment by StoyanVenDimitrov** [comment]
Hey,
is it possible to request a review from you for this pull request?

**Comment by jlibovicky** [comment]
Hi, the style looks much clsoer to what we do, you can just follow the [logs](https://travis-ci.org/ufal/neuralmonkey/jobs/246773174) from `pylint` and `pycodestyle` to fix it.

Regarding the samling, you're right that the statement on [line 427](https://github.com/ufal/neuralmonkey/blob/master/neuralmonkey/decoders/decoder.py#L427) is not used. Thanks for noticing. The place where the selection of next symbol is really happening is few lines above on [line 420](https://github.com/ufal/neuralmonkey/blob/master/neuralmonkey/decoders/decoder.py#L420). 

What I suggest you can do is:
1. make decoding loop in decoder public
2. add there an argument that will turn on sampling instead of doing `argmax`
3. call the public decoding loop with this sampling on from the traininer (this will create a new part of computational graph that will do the sampling)

You can either instantiate a decoding loop for every sample (and do sampling in parallel) which could be quite memory-demanding; or you can run the run always one only between the runs of the executable (and collect the samples one by one) which would much slower. Since MRT is meant only for fine-tuning of a pre-trained model, I would probably go for the parallel option and run the training on CPU.

**Comment by StoyanVenDimitrov** [comment]
Thanks! I actually thought, l.420 is not used in training at all. 
I'll try out what you said. 
Once I have done it with sampling, I'll also try candidates from the beam search decoder. Will what works better then.

**Comment by jlibovicky** [comment]
Please note there is a [decoder refactor](https://github.com/ufal/neuralmonkey/pull/486) going on which introduces dynamic output length which will be soon merged into master. The main difference in API is that inputs, logits, etc. will be no longer list of tensors but big tensors of dynamic length. The support is for sampling is already added to the `decoding_loop` in the refactored decoder.

**Comment by StoyanVenDimitrov** [comment]
That´s great, I see beam search is also supported. But anyway, something´s wrong with my mrt_trainer, because the memory consumption explodes with the number of samples. And I don´t know why, simply computing logits and samples didn´t seem to be that heavy. It must be something in my loss function or the way the loss is adjusted in the objective or how the gradients are later computed.

**Comment by StoyanVenDimitrov** [comment]
Do you think it is something in the mrt_los_per_sentence()?

**Comment by jindrahelcl** [comment]
Hi, unfortunately, I don't have enough time to look at your code at the moment.. I took a brief look at the `mrt_loss_per_sentence` function and I have a few remarks:

First, it seems that you are inconsistent in using logprobs and probs. You are calling a variable `logprobs`, but it stores actual probabilities (output of the softmax). You can get logprobs by using `tf.nn.log_softmax` instead. Second, if you want to use the logprobs in the subsequent computations, change `tf.pow` to multiplication, `tf.div` to subtracting, and the first `reduce_sum` to `reduce_logsumexp`. It requires a little thinking about what should be changed and what can remain though..

**Comment by StoyanVenDimitrov** [comment]
Hi Jindra,
it should be better now, thank you. But still I cannot think of the reason for this big memory consumption. For now any model with more than 5 samples crashes.

Best,
Stoyan

**Comment by StoyanVenDimitrov** [comment]
Hi,
I tried to use tensorboard to find out where the memory leak is. But it says: "Coloring by memory is only enabled if the RunMetadata proto is passed to the FileWriter when a specific session is run". Do you think there is a quick fix of this, so that memory consumption can be visualized too?

Best, 
Stoyan

**Comment by jindrahelcl** [comment]
Hi,

there is a branch called `issue_267` which is not merged yet (and is heavily out of date), in which I did the profiling. Have a look at the code, especially in the `tf_manager` module: https://github.com/ufal/neuralmonkey/blob/issue_267/neuralmonkey/tf_manager.py

J

**Comment by StoyanVenDimitrov** [comment]
Hi Jindra,
did it work back then with tensorboard? Or you didn't finish it? I did the changes and they totally make sense to me, but still see the same error in tensorboard. 

Best,
Stoyan

**Comment by StoyanVenDimitrov** [comment]
Closing this one since I work with the refactored decoder for a long time now.
